### PR TITLE
feat: support flashing prompt for focus breakpoints in the editor

### PR DESCRIPTION
**Pull request by Ricbet**
### Types

- [x] 🎉 New Features

### <samp>🤖 Generated by Copilot at 2ccbacd</samp>

*  Add less variables for debug colors and use them for editor decorations ([link](https://github.com/opensumi/core/pull/2652/files?diff=unified&w=0#diff-8478699d8b36318dea12c0dbe414766cd7abf7f534d38a9b5196b03edd358c87R1-R6), [link](https://github.com/opensumi/core/pull/2652/files?diff=unified&w=0#diff-8478699d8b36318dea12c0dbe414766cd7abf7f534d38a9b5196b03edd358c87L76-R111), [link](https://github.com/opensumi/core/pull/2652/files?diff=unified&w=0#diff-8478699d8b36318dea12c0dbe414766cd7abf7f534d38a9b5196b03edd358c87L94-R131))
*  Inject and use DebugBreakpointsService to provide events and methods for breakpoints and stack frames ([link](https://github.com/opensumi/core/pull/2652/files?diff=unified&w=0#diff-a92f8d3d8e9e5077e368d120d4405b544ea65406e1dd6c0b5fc601a9791f726bR29), [link](https://github.com/opensumi/core/pull/2652/files?diff=unified&w=0#diff-a92f8d3d8e9e5077e368d120d4405b544ea65406e1dd6c0b5fc601a9791f726bR51-R53), [link](https://github.com/opensumi/core/pull/2652/files?diff=unified&w=0#diff-b495877d18ddd7d60798cb539367bce3e910600f7d426f9f7c965e5c5a1a6e70R13), [link](https://github.com/opensumi/core/pull/2652/files?diff=unified&w=0#diff-b495877d18ddd7d60798cb539367bce3e910600f7d426f9f7c965e5c5a1a6e70R70-R72), [link](https://github.com/opensumi/core/pull/2652/files?diff=unified&w=0#diff-b495877d18ddd7d60798cb539367bce3e910600f7d426f9f7c965e5c5a1a6e70R94-R97))
*  Implement focus breakpoints stack frame line feature to highlight the stack frame that corresponds to the breakpoint that was hit ([link](https://github.com/opensumi/core/pull/2652/files?diff=unified&w=0#diff-a92f8d3d8e9e5077e368d120d4405b544ea65406e1dd6c0b5fc601a9791f726bL118-R125), [link](https://github.com/opensumi/core/pull/2652/files?diff=unified&w=0#diff-a92f8d3d8e9e5077e368d120d4405b544ea65406e1dd6c0b5fc601a9791f726bR131-R147), [link](https://github.com/opensumi/core/pull/2652/files?diff=unified&w=0#diff-a92f8d3d8e9e5077e368d120d4405b544ea65406e1dd6c0b5fc601a9791f726bL159-R182), [link](https://github.com/opensumi/core/pull/2652/files?diff=unified&w=0#diff-a92f8d3d8e9e5077e368d120d4405b544ea65406e1dd6c0b5fc601a9791f726bL166-R189), [link](https://github.com/opensumi/core/pull/2652/files?diff=unified&w=0#diff-9c4399b5585f12a56c92815dee829e9bf1ed21e94c02766795218fb952a3d707R31-R37), [link](https://github.com/opensumi/core/pull/2652/files?diff=unified&w=0#diff-5c78ed8707282f1433e7720dc171c9399cf390504427546cb332986117bd57f2L162-R171), [link](https://github.com/opensumi/core/pull/2652/files?diff=unified&w=0#diff-5c78ed8707282f1433e7720dc171c9399cf390504427546cb332986117bd57f2L176-R196))
*  Use proper types and syntax for opening resources in the editor from the breakpoints view ([link](https://github.com/opensumi/core/pull/2652/files?diff=unified&w=0#diff-5c78ed8707282f1433e7720dc171c9399cf390504427546cb332986117bd57f2L17-R19), [link](https://github.com/opensumi/core/pull/2652/files?diff=unified&w=0#diff-5c78ed8707282f1433e7720dc171c9399cf390504427546cb332986117bd57f2L176-R196))

Kapture 2023-04-26 at 15 15 14

### <samp>🤖 Generated by Copilot at 2ccbacd</samp>

This pull request enhances the debug module with new features and styles for breakpoints and stack frames. It introduces less variables for debug colors, animates the focus breakpoints stack frame line in the editor, and improves the type safety and functionality of the breakpoints view. It also refactors some code and adds new events and methods to the DebugBreakpointsService class. The changes affect the files `debug-style.less`, `debug-model.ts`, `debug-breakpoints.service.ts`, `debug-breakpoints.view.tsx`, and `debug-styles.ts`.

### CODE REVIEW

**Review by bytemain — APPROVED** [review]
LGTM

File: packages/debug/src/browser/debug-style.less [hunk 1]
```diff
@@ -1,3 +1,9 @@
+@sumi_top_stack_frame_line_color: #ffff6673;
```
Comment (erha19) [inline packages/debug/src/browser/debug-style.less:1]: 这里既然写这么长了，不如写全一点 color -> background_color , 局部变量的话，不加 @sumi_ 前缀感觉也是可以的

File: packages/debug/src/browser/editor/debug-model.ts [hunk 2]
```diff
@@ -115,13 +119,32 @@ export class DebugModel implements IDebugModel {
 
   async init() {
     const model = this.editor.getModel()!;
+    let timeOut: NodeJS.Timer;
```
Comment (erha19) [inline packages/debug/src/browser/editor/debug-model.ts:122]: 前端代码里面，这里的类型不太对吧，而且一般命名为 timer 多一点，timeOut 很怪

File: packages/debug/src/browser/editor/debug-model.ts [hunk 3]
```diff
@@ -115,13 +119,32 @@ export class DebugModel implements IDebugModel {
...
+            options: options.FOCUS_BREAKPOINTS_STACK_FRAME_DECORATION,
+            range,
+          },
+        ]);
+
+        if (timeOut) {
+          clearTimeout(timeOut);
+        }
+
+        timeOut = global.setTimeout(() => {
```
Comment (erha19) [inline packages/debug/src/browser/editor/debug-model.ts:143]: 只会在前端运行, 这里的 global 可以省略?

File: packages/debug/src/browser/editor/debug-model.ts (hunk 2 above)
Comment (Ricbet) [reply packages/debug/src/browser/editor/debug-model.ts:122]: 这里是为了过类型检查，不过确实不应该用 NodeJS.Timer

File: packages/debug/src/browser/debug-style.less (hunk 1 above)
Comment (Ricbet) [reply packages/debug/src/browser/debug-style.less:1]: 就叫 top_stack_frame_line_bg_color 吧

**Comment by Ricbet** [comment]
- [x] 新增 preference 配置项，控制是否开启该功能，默认开启

File: packages/debug/src/browser/editor/debug-model.ts [hunk 4]
```diff
@@ -115,13 +122,38 @@ export class DebugModel implements IDebugModel {
...
+        }
+
+        this.renderFrames([
+          {
+            options: options.FOCUS_BREAKPOINTS_STACK_FRAME_DECORATION,
+            range,
+          },
+        ]);
+
+        if (timmer) {
```
Comment (erha19) [inline packages/debug/src/browser/editor/debug-model.ts:148]: typo, timmer -> timer

**Review by erha19 — CHANGES_REQUESTED** [review]

**Review by erha19 — APPROVED** [review]

**Pull request merged**